Add osu! mod "Approach Different"

peppy opened the pull request:
Taking over from https://github.com/ppy/osu/pull/8284/ as I can't push to the branch there. Credit to @BlauFx for the idea / initial implementation.

Have renamed and tidied up the easing naming and mapping logic.


https://user-images.githubusercontent.com/191335/122597130-e7149880-d0a5-11eb-9853-458a3f3f6fa3.mp4

BlauFx committed:
Add approachcircle mod

BlauFx committed:
Change label of attribute

BlauFx committed:
Remove empty lines

BlauFx committed:
Remove redundant check

Co-Authored-By: Bartłomiej Dach <dach.bartlomiej@gmail.com>

BlauFx committed:
Fix bdach's suggestion

BlauFx committed:
Resolving bdach's requested to change the enum member names and reduce the amount of selectable easing types

peppy committed:
Rename mod and fix easing mappings / naming

peppy committed:
Merge branch 'master' into AddApproachCircleMod

jvyden commented on osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs:
```diff
@@ -0,0 +1,110 @@
+        public override string Acronym => "AD";
+        public override string Description => "Never trust the approach circles...";
+        public override double ScoreMultiplier => 1;
+        public override IconUsage? Icon { get; } = FontAwesome.Regular.Circle;
```
```suggestion
        public override IconUsage? Icon { get; } = FontAwesome.Regular.QuestionCircle;
```

peppy replied:
No thanks

jvyden commented on osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs:
```diff
@@ -0,0 +1,110 @@
+
+namespace osu.Game.Rulesets.Osu.Mods
+{
+    internal class OsuModDifferentApproach : Mod, IApplicableToDrawableHitObjects
```
This should be changed to `OsuModApproachDifferent`.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs:
```diff
@@ -0,0 +1,110 @@
+        public override double ScoreMultiplier => 1;
+        public override IconUsage? Icon { get; } = FontAwesome.Regular.Circle;
+
+        [SettingSource("Easing", "Change the animation curve of the approach circles.", 0)]
```
This reads worse.

jvyden replied:
> This reads worse.

Fair enough for the description, but the term `Easing` doesn't really sit right in my eyes, at least from a player's point of view.

peppy replied:
Oh right. That should be changed to "style".

jvyden replied:
Ah, that's the word I was looking for. I'll update the suggestion.

peppy committed:
Rename class to match new name

jvyden commented on osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs:
```diff
@@ -0,0 +1,110 @@
+        public override double ScoreMultiplier => 1;
+        public override IconUsage? Icon { get; } = FontAwesome.Regular.Circle;
+
+        [SettingSource("Easing", "Change the animation curve of the approach circles.", 0)]
```
```suggestion
        [SettingSource("Style", "Change the animation curve of the approach circles.", 0)]
```

peppy committed:
Rename "easing" references to be "style" instead

ekrctb commented on osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs:
```diff
@@ -0,0 +1,110 @@
+
+                    var obj = hitCircle.HitObject;
+
+                    hitCircle.BeginAbsoluteSequence(obj.StartTime - obj.TimePreempt, true);
```
Shouldn't `BeginAbsoluteSequence` be use with `using`?

ekrctb commented on osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs:
```diff
@@ -0,0 +1,110 @@
+            });
+        }
+
+        private Easing getEasing(AnimationStyle approachEasing)
```
Nit: the parameter name is still using "easing" while the type name is renamed to "style".

ekrctb commented on osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs:
```diff
@@ -0,0 +1,110 @@
+
+        public enum AnimationStyle
+        {
+            Default,
```
I wonder the mod should use an animation style other than the `Easing.None` by default.
A user will see no difference to the gameplay when the mod is enabled until the user finds the customization menu.

peppy replied:
Fair call. I'm not sure if we even want default on here..

ekrctb commented:
(Not specific to this PR)
In 16∶9 resolution, the mod button is not clickable when the customization menu is opening because the menu is overlapping the mod button.

peppy committed:
Remove default (and make default "Gravity")

peppy committed:
Simplify transform logic

peppy committed:
Change order of settings to make scrolling easier

There's an issue with dropdown menus nested inside a scroll view being
very frustrating to scroll to off-screen items. This works around that
to some extent by giving the user more "parent-scrollable" space to
mouse wheel or drag over.

BlauFx commented:
Thanks @peppy for picking this one up, I completely forgot about it. I'm glad to see, that this mod made it's way into osu!.
Thank you
